services/graph/pkg/identity/ldap.go

```diff
@@ -470,6 +473,68 @@ func (i *LDAP) GetUsers(ctx context.Context, oreq *godata.GoDataRequest) ([]*lib
 	return users, nil
 }
 
+func (i *LDAP) changeUserName(ctx context.Context, dn, originalUserName, newUserName string) (*ldap.Entry, error) {
+	logger := i.logger.SubloggerWithRequestID(ctx)
+	groups, err := i.getGroupsForUser(dn)
+	if err != nil {
+		return nil, err
+	}
+
+	newDN := fmt.Sprintf("%s=%s", i.userAttributeMap.userName, newUserName)
+
+	logger.Debug().Str("originalDN", dn).Str("newDN", newDN).Msg("Modifying DN")
+	mrdn := ldap.NewModifyDNRequest(dn, newDN, true, "")
+
+	if err := i.conn.ModifyDN(mrdn); err != nil {
+		var lerr *ldap.Error
+		logger.Debug().Str("originalDN", dn).Str("newDN", newDN).Err(err).Msg("Failed to modify DN")
+		if errors.As(err, &lerr) {
+			if lerr.ResultCode == ldap.LDAPResultEntryAlreadyExists {
+				err = errorcode.New(errorcode.NameAlreadyExists, lerr.Error())
+			}
+		}
+		return nil, err
+	}
+
+	u, err := i.getUserByDN(fmt.Sprintf("%s,%s", newDN, i.userBaseDN))
+	if err != nil {
+		return nil, err
+	}
+
+	for _, g := range groups {
+		err = i.renameMemberInGroup(ctx, g, dn, u.DN)
+		// This could leave the groups in an inconsistent state, might be a good idea to
+		// add a defer that changes everything back on error. Ideally, this entire function
+		// should be atomic, but LDAP doesn't support that.
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return u, nil
+}
+
+func (i *LDAP) renameMemberInGroup(ctx context.Context, group *ldap.Entry, oldMember, newMember string) error {
+	logger := i.logger.SubloggerWithRequestID(ctx)
+	logger.Debug().Str("oldMember", oldMember).Str("newMember", newMember).Msg("replacing group member")
+	members := group.GetEqualFoldAttributeValues(i.groupAttributeMap.member)
+	match := -1
+	for i, m := range members {
+		if m == oldMember {
+			match = i
+		}
+	}
+	if match != -1 {
+		members[match] = newMember
+		mr := ldap.NewModifyRequest(group.DN, nil)
+		mr.Replace(i.groupAttributeMap.member, members)
```

Comment:
This is a bit of a nitpick, I'll do it anyway ;-)

LDAP wise I think it's more effecient/correct do a Delete/Add Modify Operation here instead of the Replace. It should also simplify the code a bit (you wouldn't need to fiddle with the members slice, e.g.)

```
mr.Delete((i.groupAttributeMap.member, []string{oldMember})
mr.Add(i.groupAttributeMap.member, []string{newMember})
```

If you do this within a single LDAPModify Request it even is atomic.

Reply:
Thanks, I thought this would remove all the other members, so that's why I didn't do that, but I changed it now and it works nicely with less code. :)